### ISSUE TITLE
Camunda Exporter is not able to consume treePath

### ISSUE THREAD

**Issue by ChrisKujawa** [camunda/camunda#26048]
**Describe the bug**

When running a benchmark we see endless [logs](https://console.cloud.google.com/logs/query;query=resource.type%3D%22k8s_container%22%0Aresource.labels.project_id%3D%22zeebe-io%22%0Aresource.labels.location%3D%22europe-west1-b%22%0Aresource.labels.cluster_name%3D%22zeebe-cluster%22%0Aresource.labels.namespace_name%3D%22ck-main-exporter%22%0Alabels.k8s-pod%2Fapp%3D%22camunda-platform%22%0Alabels.k8s-pod%2Fapp_kubernetes_io%2Fcomponent%3D%22zeebe-broker%22%0Alabels.k8s-pod%2Fapp_kubernetes_io%2Finstance%3D%22zeebe-benchmark-test%22%0Alabels.k8s-pod%2Fapp_kubernetes_io%2Fmanaged-by%3D%22Helm%22%0Alabels.k8s-pod%2Fapp_kubernetes_io%2Fname%3D%22camunda-platform%22%0Alabels.k8s-pod%2Fapp_kubernetes_io%2Fpart-of%3D%22camunda-platform%22;pinnedLogId=2024-12-12T21:20:50.685776443Z%2F6nksld3qseaqfa3z;cursorTimestamp=2024-12-12T21:20:50.691136797Z;duration=PT1H?inv=1&invt=Abj9IQ&project=zeebe-io) of:

```
No elementInstancePath is provided for flow node instance id: 6755399442562079. TreePath will be set to default value.
```

**Comment by ChrisKujawa** [comment]
https://zeebe-io.github.io/zeebe-chaos/2024/12/12/News-from-Camunda-Exporter-project

**Comment by sdorokhova** [comment]
16.12.2024:
I researched the issue:
1. Locally on current `main` codebase was not reproducible.
2. Several clusters of version `8.7.0-alpha2` shows the problem. With enabled debug logging (ZEEBE_DEBUG=true), we see that Zeebe records don't provide `elementInstancePath`, `processDefinitionPath` and `callingElementPath` fields:
```
message: "{"partitionId":1,"value":{"flowScopeKey":2251799813697539,"bpmnEventType":"NONE","parentElementInstanceKey":-1,"bpmnElementType":"START_EVENT","version":1,"tenantId":"<default>","elementId":"Event_18mj23t","bpmnProcessId":"call-activity-process","parentProcessInstanceKey":-1,"processInstanceKey":2251799813697535,"processDefinitionKey":2251799813686719},"key":2251799813697540,"timestamp":1734363222221,"valueType":"PROCESS_INSTANCE","recordType":"EVENT","rejectionType":"NULL_VAL","rejectionReason":"","recordVersion":1,"sourceRecordPosition":24935,"authorizations":{"authorized_tenants":["<default>"]},"intent":"ELEMENT_ACTIVATING","operationReference":-1,"brokerVersion":"8.7.0","position":24951}"
serviceContext: {
service: "zeebe"
version: "8.7.0-alpha2"
}
```
3. On clusters of SNAPSHOT version not reproducible. Operate UI renders data properly. Debug logging shows fields properly filled:
```
message: "{"partitionId":1,"value":{"bpmnEventType":"NONE","parentElementInstanceKey":-1,"elementInstancePath":[[2251799813693044,2251799813693048,2251799813693049]],"processDefinitionPath":[2251799813685731],"callingElementPath":[],"bpmnElementType":"START_EVENT","flowScopeKey":2251799813693048,"version":1,"tenantId":"<default>","bpmnProcessId":"call-activity-process","processInstanceKey":2251799813693044,"processDefinitionKey":2251799813685731,"parentProcessInstanceKey":-1,"elementId":"Event_18mj23t"},"key":2251799813693049,"timestamp":1734364226830,"valueType":"PROCESS_INSTANCE","authorizations":{"authorized_tenants":["<default>"]},"intent":"ELEMENT_ACTIVATING","brokerVersion":"8.7.0","operationReference":-1,"recordType":"EVENT","sourceRecordPosition":15619,"rejectionType":"NULL_VAL","rejectionReason":"","recordVersion":1,"position":15635}"
serviceContext: {
service: "zeebe"
version: "8.7.0-alpha2-test-treepath"
}
```

My assumption that something was fixed in between alpha2 build and today. I propose to observe further and make sure to test this in alpha3 release.

@ArezooAzar could we plan testing `treePath` related test cases within alpha3 release test? I'm ready to sync with you to clarify which tests are this.

**Comment by sdorokhova** [comment]
Example of right and wrong `intra treePath` how is it visible in UI:
![image](https://github.com/user-attachments/assets/a91e5db8-234f-4621-8acc-424b6095c4b8)
![image (1)](https://github.com/user-attachments/assets/fad0e9b1-2f2a-4892-b222-ef498b9ae5d3)

**Comment by sdorokhova** [comment]
These test cases will be included in 8.7.0-alpha3 release test plan:
https://camunda.testrail.com/index.php?/cases/view/2845
https://camunda.testrail.com/index.php?/cases/view/3644
https://camunda.testrail.com/index.php?/cases/view/3864
https://camunda.testrail.com/index.php?/cases/view/3672
https://camunda.testrail.com/index.php?/cases/view/550947

**Comment by ChrisKujawa** [comment]
@mustafadagher might be related to https://github.com/camunda/camunda/issues/26755

**Comment by mustafadagher** [comment]
@ChrisKujawa I assume from @sdorokhova that this bug has not been reproducible for versions after `8.7.0-alpha2`. Can we assume it has been fixed and reopen again if ever encountered?

**Comment by ChrisKujawa** [comment]
Yes

**Comment by lenaschoenburg** [comment]
I was just testing the latest version (docker image camunda/camunda:SNAPSHOT) and I'm seeing the same warning logged over an over: https://cloudlogging.app.goo.gl/jKDAjEzNWcY4r2GD9

**Comment by ChrisKujawa** [comment]
@mustafadagher can you please look at this.

In the benchmarks we btw dont use user task listeners

**Comment by ChrisKujawa** [comment]
> Is there any way to find out what was the BPMN Element Type of some of these elements with no treepath?
> Digging a little bit, I found out that for example we do mark `Compensation Events` as `ACTIVATING` without calculating the treepath for it. 
> https://github.com/camunda/camunda/blob/5486adf334656dcf1c32721a0b550dfdbf16e84c/zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnCompensationSubscriptionBehaviour.java#L294-L295
> Maybe there are other cases, but just going through every usage in the code base could be tedious.
> 


What i think you could do is adding some assert in some event handlers like activating and then run the test suite and check where it fails.

Alternative add logs and run a some tests or even a benchmark to better trace it.

It is a simple process with one service task so it should be easily reproducible.

@lenaschoenburg or was this in the realistic benchmarks?



> ---
> **UPDATE:**
> I found that calculating the treepath is also missing in ElementActivationBehavior (which is used during ProcessInstance creation and Modification)
> 
> https://github.com/camunda/camunda/blob/5486adf334656dcf1c32721a0b550dfdbf16e84c/zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/common/ElementActivationBehavior.java#L369-L385

Great!

**Comment by lenaschoenburg** [comment]
This was on the realistic workload. The simple one-task process doesn't have this problem.

**Comment by mustafadagher** [comment]
I extended the `ModifyProcessInstanceTest` locally with the assertion below aiming to prove a different issue (where activation with modification doesn't include tree path, turns out it does), but I ended up finding out that `SUB_PROCESS` and `EVENT_SUB_PROCESS` element instances to not have a treepath set. 

in the end of `verifyThatElementIsActivated`, I added the following (which is not 100% accurate for all cases, but good enough to prove the idea)

```java
elementInstanceEvents.stream()
        .filter(e -> e.getIntent().equals(ProcessInstanceIntent.ELEMENT_ACTIVATING))
        .forEach(
            elementActivatingEvent -> {
              Assertions.assertThat(elementActivatingEvent)
                  .describedAs(
                      "Expect the element instance activating event to contain the element id")
                  .isNotNull()
                  .extracting(Record::getValue)
                  .extracting(
                      ProcessInstanceRecordValue::getElementInstancePath,
                      ProcessInstanceRecordValue::getProcessDefinitionPath)
                  .containsOnly(
                      List.of(
                          List.of(
                              processActivatedRecord.getKey(), elementActivatingEvent.getKey())),
                      List.of(processActivatedEvent.getProcessDefinitionKey()));
            });
```

You need to make this change in the beginning of the method too for that to compile 
```java
 final Record<ProcessInstanceRecordValue> processActivatedRecord =
        RecordingExporter.processInstanceRecords(ProcessInstanceIntent.ELEMENT_ACTIVATED)
            .withProcessInstanceKey(processInstanceKey)
            .withElementType(BpmnElementType.PROCESS)
            .getFirst();
    final var processActivatedEvent = processActivatedRecord.getValue();
```

**Comment by mustafadagher** [comment]
@ChrisKujawa I added a few tests [here](https://github.com/camunda/camunda/pull/28425), only one of them fails, which is the one trying to verify that treepath exists for `Compensation BoundaryEvents`, which assures my doubt in [this comment](https://github.com/camunda/camunda/issues/26048#issuecomment-2668230055)

**Comment by mustafadagher** [comment]
> But compensation is not used in the processes we use in our benchnark 🤔

yes, I think there are more types to catch. I'm working on extending the randomized properties test as you suggested. I have a feeling that `BOUNDARY_EVENT`s in general are an issue, but I'm yet to find out
